### DISCUSSION TITLE
How to read in VCF files for use in tsinfer?

### DISCUSSION THREAD

**Discussion by hyanwong** [tskit-dev/tsinfer#524] [Q&A] [answered]
How do I import a VCF file into tsinfer?

**Comment by hyanwong** [comment, marked as answer]
Replying to my own question, there's code provided in the [tutorial](https://tsinfer.readthedocs.io/en/latest/tutorial.html#reading-a-vcf) for reading VCF files, including reading in some of the metadata in the VCF (details further down that page). If you have a very large VCF file and it is taking a long time, then it is possible to read chunks of your VCF in parallel: see https://github.com/tskit-dev/tsinfer/issues/277#issuecomment-652024871 for example code.

Note that in the longer term, we may be moving tsinfer to using [sgkit](https://pystatgen.github.io/sgkit/latest/) as an import framework: when this happens this answer may become obsolete.

Also note that you if you are importing VCFs into tsinfer, you may wish to investigate [how to infer ancestral states properly](https://github.com/tskit-dev/tsinfer/discussions/523).

**Comment by npb596** [comment]
Hi Yan,

Have there been any updates to this? I ran a script with the exact code given in that tutorial link (except with file path changed to my VCF) on a VCF generated by SLiM (where I manually input a sequence length in header) and receive this error:

/mnt/apps/users/nbailey/conda/lib/python3.12/site-packages/tsinfer/formats.py:458: FutureWarning: The LMDBStore is deprecated and will be removed in a Zarr-Python version 3, see https://github.com/zarr-developers/zarr-python/issues/1274 for more information.
  return zarr.LMDBStore(self.path, subdir=False, map_size=map_size)
Traceback (most recent call last):
  File "/mnt/shared/projects/uosa/Nick_Bailey/vcf_tskit.py", line 51, in <module>
    with tsinfer.SampleData(
         ^^^^^^^^^^^^^^^^^^^
  File "/mnt/apps/users/nbailey/conda/lib/python3.12/site-packages/tsinfer/formats.py", line 952, in __init__
    super().__init__(**kwargs)
  File "/mnt/apps/users/nbailey/conda/lib/python3.12/site-packages/tsinfer/formats.py", line 384, in __init__
    store = self._new_lmdb_store(max_file_size)
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/mnt/apps/users/nbailey/conda/lib/python3.12/site-packages/tsinfer/formats.py", line 458, in _new_lmdb_store
    return zarr.LMDBStore(self.path, subdir=False, map_size=map_size)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/mnt/apps/users/nbailey/conda/lib/python3.12/site-packages/zarr/storage.py", line 2331, in __init__
    self.db = lmdb.open(path, **kwargs)
              ^^^^^^^^^^^^^^^^^^^^^^^^^
lmdb.MemoryError: /mnt/shared/projects/uosa/Nick_Bailey/BS_1_p1.samples: Cannot allocate memory

I can provide the exact script and VCF if needed but again both shouldn't have any substantive deviation from what's assumed in tutorial.

Cheers,
Nick

**Reply by hyanwong** [reply]
There will be a new release of tsinfer in the next week or so describing the VariantData format, which can be used to read in VCFs via vcf2zarr. It should be available as a pre-release (v 0.4) now though, and documented in the "latest" (not "stable") docs.

Note that there was also a memory allocation issue in previous versions of running under Windows. Is that the case for you?

**Reply by npb596** [reply]
Thank you, I'll give the pre-release a look and await the new release!

As for the memory issue I'm running on a Linux cluster so no Windows involved, so not sure on the cause of the MemoryError I'm getting.

**Reply by benjeffery** [reply]
What filesystem is `/mnt/shared`? When opening the lmdb file `mmap` is used with a large initial size. This isn't an issue on most linux filesystems, but some network mounts could have a limit lower than that used by `tsinfer` (2**40). You can specify a lower size by using the `max_file_size` argument to `SampleData`.

**Reply by npb596** [reply]
I believe the file system is BeeGFS (info from my cluster: https://help.cropdiversity.ac.uk/data-storage.html#beegfs-storage) though if that response doesn't make sense pardon my ignorance of what a file system refers to here. In any case, using 2**30 for max_file_size definitely seems to fix that particular error, so thank you. I now realize a 1TB BS_1_p1.samples and a smaller BS_1_p1.samples-lock file were being generated and now the former is 1GB, so I guess that makes sense. The new error, which I'm actively troubleshooting, is:

/mnt/apps/users/nbailey/conda/lib/python3.12/site-packages/tsinfer/formats.py:458: FutureWarning: The LMDBStore is deprecated and will be removed in a Zarr-Python version 3, see https://github.com/zarr-developers/zarr-python/issues/1274 for more information.
  return zarr.LMDBStore(self.path, subdir=False, map_size=map_size)
Traceback (most recent call last):
  File "/mnt/shared/projects/uosa/Nick_Bailey/vcf_tskit.py", line 54, in <module>
    add_diploid_sites(vcf, samples)
  File "/mnt/shared/projects/uosa/Nick_Bailey/vcf_tskit.py", line 12, in add_diploid_sites
    progressbar = tqdm(total=samples.sequence_length, desc="Read VCF", unit='bp')
                  ^^^^
NameError: name 'tqdm' is not defined

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/mnt/shared/projects/uosa/Nick_Bailey/vcf_tskit.py", line 51, in <module>
    with tsinfer.SampleData(
         ^^^^^^^^^^^^^^^^^^^
  File "/mnt/apps/users/nbailey/conda/lib/python3.12/site-packages/tsinfer/formats.py", line 414, in __exit__
    self.finalise()
  File "/mnt/apps/users/nbailey/conda/lib/python3.12/site-packages/tsinfer/formats.py", line 1915, in finalise
    raise ValueError("Must add at least one sample individual")
ValueError: Must add at least one sample individual

**Reply by benjeffery** [reply]
Add `import tqdm` to the top of your script.

**Reply by npb596** [reply]
That gave me this error:

TypeError: 'module' object is not callable. Did you mean: 'tqdm.tqdm(...)'?

and then I decided I don't really need/want a progressbar so I commented out the relevant lines. I now have this:

/mnt/apps/users/nbailey/conda/lib/python3.12/site-packages/tsinfer/formats.py:458: FutureWarning: The LMDBStore is deprecated and will be removed in a Zarr-Python version 3, see https://github.com/zarr-developers/zarr-python/issues/1274 for more information.
  return zarr.LMDBStore(self.path, subdir=False, map_size=map_size)
/mnt/apps/users/nbailey/conda/lib/python3.12/site-packages/tsinfer/formats.py:428: FutureWarning: The LMDBStore is deprecated and will be removed in a Zarr-Python version 3, see https://github.com/zarr-developers/zarr-python/issues/1274 for more information.
  store = zarr.LMDBStore(
Traceback (most recent call last):
  File "/mnt/shared/projects/uosa/Nick_Bailey/vcf_tskit.py", line 54, in <module>
    add_diploid_sites(vcf, samples)
  File "/mnt/shared/projects/uosa/Nick_Bailey/vcf_tskit.py", line 39, in add_diploid_sites
    samples.add_site(pos, genotypes, alleles, ancestral_allele=ancestral_allele)
  File "/mnt/apps/users/nbailey/conda/lib/python3.12/site-packages/tsinfer/formats.py", line 1832, in add_site
    raise ValueError("Alleles must be distinct")
ValueError: Alleles must be distinct

**Reply by benjeffery** [reply]
Ah, sorry should have been `from tqdm import tqdm`.

That error means your alleles array for the given site has repeats in. Each allele needs to be unique.

**Reply by npb596** [reply]
Hmm, looking at the VCF I have a guess what's going on there. Some sites have the same ALT allele repeated twice like the A here (in order of CHROM, POS, ID, REF ALT):

1       892     .       T       A,A     1000    PASS

I guess it had to do with how SLiM generates the files. I could filter these out, does that seem like the right idea?

**Reply by npb596** [reply]
Yes that seemed to be the issue, along with other similarly odd sites. A clunky but viable solution right now is:

```cat BS_1_p1.vcf | grep -v -E "A\s*\sA" | grep -v -E "C\s*\sC" | grep -v -E "G\s*\sG" | grep -v -E "T\s*\sT" | grep -v "[ACGT],[ACGT]" > test.vcf```

where the first four grep conditions exclude sites where the ref and alt are the same allele and the final one excludes all multi-allelic sites, because it wasn't just the multi-allelic sites with the same two alt alleles that caused the issue. The output after that seems to be a valid samples file as per the tutorial. Thanks for all the quick assistance!
